File: src/main/kotlin/com/vitiquest/peerreview/jira/JiraIntegrationService.kt
package com.vitiquest.peerreview.jira

import com.vitiquest.peerreview.bitbucket.PullRequest
import com.vitiquest.peerreview.settings.PluginSettings

enum class ReviewOutcome {
    APPROVED,
    MERGED,
    DECLINED
}

enum class JiraSyncStatus {
    UPDATED,
    SKIPPED_NOT_CONFIGURED,
    SKIPPED_NO_ISSUE
}

data class JiraSyncResult(
    val status: JiraSyncStatus,
    val issueKey: String? = null,
    val reassigned: Boolean = false,
    val assigneeLabel: String? = null
) {
    fun userMessage(): String? = when (status) {
        JiraSyncStatus.UPDATED -> buildString {
            append("JIRA")
            issueKey?.let { append(" $it") }
            append(" updated")
            if (reassigned && !assigneeLabel.isNullOrBlank()) {
                append(" and reassigned to $assigneeLabel")
            }
            append(".")
        }
        else -> null
    }
}

class JiraIntegrationService {

    fun syncReviewOutcome(pr: PullRequest, outcome: ReviewOutcome, summary: String? = null): JiraSyncResult {
        val settings = PluginSettings.instance
        val baseUrl = settings.getJiraBaseUrl()
        val email = settings.getJiraEmail()
        val token = settings.getJiraApiToken()

        if (baseUrl.isBlank() || email.isBlank() || token.isBlank()) {
            return JiraSyncResult(JiraSyncStatus.SKIPPED_NOT_CONFIGURED)
        }

        val issueKey = findAttachedIssueKey(pr, settings.getJiraIssueKeyPattern())
            ?: return JiraSyncResult(JiraSyncStatus.SKIPPED_NO_ISSUE)

        val client = JiraClient(baseUrl, email, token)
        client.getIssue(issueKey)

        val commentBody = when (outcome) {
            ReviewOutcome.APPROVED,
            ReviewOutcome.MERGED -> "Code Review Passed."
            ReviewOutcome.DECLINED -> summary?.trim().takeUnless { it.isNullOrBlank() }
                ?: "PR was declined. AI summary was unavailable."
        }
        client.addPlainTextComment(issueKey, commentBody)

        if (outcome != ReviewOutcome.DECLINED) {
            return JiraSyncResult(JiraSyncStatus.UPDATED, issueKey = issueKey)
        }

        val assignee = resolvePrAuthor(pr, client)
        if (assignee != null) {
            client.assignIssue(issueKey, assignee.accountId)
        }

        return JiraSyncResult(
            status = JiraSyncStatus.UPDATED,
            issueKey = issueKey,
            reassigned = assignee != null,
            assigneeLabel = assignee?.displayName?.ifBlank { pr.author.displayName }
        )
    }

    private fun resolvePrAuthor(pr: PullRequest, client: JiraClient): JiraUser? {
        val searchTerms = listOf(pr.author.nickname, pr.author.displayName)
            .map { it.trim() }
            .filter { it.isNotBlank() }
            .distinct()

        searchTerms.forEach { term ->
            val matches = client.searchUsers(term)
                .filter { it.active }

            matches.firstOrNull {
                it.displayName.equals(term, ignoreCase = true) ||
                    (it.emailAddress?.equals(term, ignoreCase = true) == true)
            }?.let { return it }

            if (matches.size == 1) {
                return matches.first()
            }
        }

        return null
    }

    private fun findAttachedIssueKey(pr: PullRequest, configuredPattern: String): String? {
        val pattern = runCatching {
            Regex(
                configuredPattern.ifBlank { "[A-Z][A-Z0-9]+-\\d+" },
                setOf(RegexOption.IGNORE_CASE)
            )
        }.getOrElse {
            Regex("[A-Z][A-Z0-9]+-\\d+", setOf(RegexOption.IGNORE_CASE))
        }

        val sources = listOf(
            pr.title,
            pr.description,
            pr.source.branch.name,
            pr.destination.branch.name,
            pr.links.html.href
        )

        return sources.asSequence()
            .mapNotNull { source -> pattern.find(source)?.value }
            .map { it.uppercase() }
            .firstOrNull()
    }
}

File: vscode-plugin/src/jira/JiraIntegrationService.ts
import { PullRequest } from '../models/PullRequest';
import { Settings } from '../settings/Settings';
import { JiraClient, JiraUser } from './JiraClient';
import { Logger } from '../utils/Logger';

export type ReviewOutcome = 'APPROVED' | 'MERGED' | 'DECLINED' | 'CHANGES_REQUESTED';

export type JiraSyncStatus = 'UPDATED' | 'SKIPPED_NOT_CONFIGURED' | 'SKIPPED_NO_ISSUE';

export interface JiraSyncResult {
  status: JiraSyncStatus;
  issueKey?: string;
  reassigned?: boolean;
  assigneeLabel?: string;
}

export function jiraSyncMessage(result: JiraSyncResult): string | undefined {
  if (result.status !== 'UPDATED') return undefined;
  let msg = 'JIRA';
  if (result.issueKey) msg += ` ${result.issueKey}`;
  msg += ' updated';
  if (result.reassigned && result.assigneeLabel) {
    msg += ` and reassigned to ${result.assigneeLabel}`;
  }
  return msg + '.';
}

/**
 * Syncs PR review outcomes to JIRA Cloud.
 * Detects the issue key from the PR title, description, or branch name.
 */
export class JiraIntegrationService {
  async syncReviewOutcome(
    pr: PullRequest,
    outcome: ReviewOutcome,
    summary?: string
  ): Promise<JiraSyncResult> {
    const settings = Settings.instance;
    const baseUrl = settings.jiraBaseUrl;
    const email = settings.jiraEmail;
    const token = await settings.getJiraApiToken();

    if (!baseUrl || !email || !token) {
      return { status: 'SKIPPED_NOT_CONFIGURED' };
    }

    const issueKey = findAttachedIssueKey(pr, settings.jiraIssueKeyPattern);
    if (!issueKey) {
      Logger.info(`[JIRA] No issue key found in PR #${pr.id} (title="${pr.title}", branch="${pr.source.branch.name}") — skipping sync.`);
      return { status: 'SKIPPED_NO_ISSUE' };
    }

    Logger.info(`[JIRA] Syncing issue ${issueKey} | PR #${pr.id} | outcome=${outcome}`);

    const client = new JiraClient(baseUrl, email, token);

    await client.getIssue(issueKey);
    await client.addPlainTextComment(issueKey, buildCommentBody(outcome, summary));

    if (outcome === 'APPROVED' || outcome === 'MERGED') {
      Logger.info(`[JIRA] ${issueKey} comment posted (${outcome}).`);
      return { status: 'UPDATED', issueKey };
    }

    // Attempt to assign to PR author for DECLINED / CHANGES_REQUESTED
    const assignee = await resolvePrAuthor(pr, client);
    if (assignee) {
      await client.assignIssue(issueKey, assignee.accountId);
      Logger.info(`[JIRA] ${issueKey} comment posted and reassigned to ${assignee.displayName}.`);
    } else {
      Logger.warn(`[JIRA] ${issueKey} comment posted but could not resolve JIRA user for "${pr.author.displayName}" — not reassigned.`);
    }

    return {
      status: 'UPDATED',
      issueKey,
      reassigned: !!assignee,
      assigneeLabel: assignee?.displayName ?? pr.author.displayName,
    };
  }
}

function buildCommentBody(outcome: ReviewOutcome, summary?: string): string {
  switch (outcome) {
    case 'APPROVED':
    case 'MERGED':
      return 'Code Review Passed.';
    case 'DECLINED':
      return summary?.trim() || 'PR was declined. AI summary was unavailable.';
    case 'CHANGES_REQUESTED':
      return summary?.trim() || 'Changes requested on PR. AI summary was unavailable.';
  }
}

async function resolvePrAuthor(pr: PullRequest, client: JiraClient): Promise<JiraUser | undefined> {
  const searchTerms = [pr.author.nickname, pr.author.displayName]
    .map((t) => t.trim())
    .filter((t) => t.length > 0)
    .filter((t, i, arr) => arr.indexOf(t) === i);

  for (const term of searchTerms) {
    try {
      const matches = (await client.searchUsers(term)).filter((u) => u.active);

      const exact = matches.find(
        (u) =>
          u.displayName.toLowerCase() === term.toLowerCase() ||
          u.emailAddress?.toLowerCase() === term.toLowerCase()
      );
      if (exact) return exact;

      if (matches.length === 1) return matches[0];
    } catch {
      // ignore per-term search failures
    }
  }
  return undefined;
}

function findAttachedIssueKey(pr: PullRequest, configuredPattern: string): string | undefined {
  let pattern: RegExp;
  try {
    pattern = new RegExp(configuredPattern || '[A-Z][A-Z0-9]+-\\d+', 'i');
  } catch {
    pattern = /[A-Z][A-Z0-9]+-\d+/i;
  }

  const sources = [
    pr.title,
    pr.description,
    pr.source.branch.name,
    pr.destination.branch.name,
    pr.links.html.href,
  ];

  for (const src of sources) {
    const match = pattern.exec(src);
    if (match) return match[0].toUpperCase();
  }
  return undefined;
}
